sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/EnumTrailingCommaCheck.java

```diff
@@ -0,0 +1,159 @@
+ * }
+ * </pre>
+ *
+ * <p>However, you could also add a semicolon behind that comma on the same line, which would raise
+ * a violation
+ *
+ * <pre>
+ * enum Type {
+ *     ALPHA,
+ *     BETA,
+ *     GAMMA,
+ *     DELTA,; // violation
+ * }
+ * </pre>
+ * In this case the semicolon should be removed. However, if there is more in the enum body, the
+ * semicolon should be placed on a line by itself.
+ *
+ * <p>An example of how to configure the check is:
+ * <pre>
+ * &lt;module name="EnumTrailingComma"/&gt;
+ * </pre>
+ *
+ * <p>Please note that using this check together with {@code NoWhitespaceBefore} or
+ * {@code SeparatorWrap} may create conflicts with enums that contain a body:
+ * {@code EnumTrailingComma} enforces the semicolon on a separate line while
+ * {@code NoWhiteSpaceBefore} does not allow the semicolon to be preceded with whitespace and
+ * {@code SeparatorWrap} expects the semicolon to be on the same line as the last enum constant.
+ *
+ * @author <a href="kariem.hussein@gmail.com">Kariem Hussein</a>
+ */
+public class EnumTrailingCommaCheck extends AbstractCheck {
+
+    /** Key for warning message text in "messages.properties" file. */
+    public static final String MSG_KEY = "enum.trailing.comma";
+    /** Key for warning message text in "messages.properties" file. */
+    public static final String MSG_KEY_SEMI = "enum.trailing.comma.semi";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return new int[] {TokenTypes.ENUM_DEF};
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public void visitToken(DetailAST enumDef) {
+        final DetailAST enumConstBlock = enumDef.findFirstToken(TokenTypes.OBJBLOCK);
+
+        final DetailAST enumConstLeft = enumConstBlock.findFirstToken(TokenTypes.LCURLY);
+        final DetailAST enumConstRight = enumConstBlock.findFirstToken(TokenTypes.RCURLY);
+
+        // Only check, if block is multi-line and there are more than one enum constants
+        if (enumConstLeft.getLineNo() != enumConstRight.getLineNo()
+                && enumConstBlock.getChildCount(TokenTypes.ENUM_CONSTANT_DEF) > 1) {
+            final DetailAST constant = enumConstBlock.findFirstToken(TokenTypes.ENUM_CONSTANT_DEF);
+            final DetailAST lastComma = getLastComma(constant);
+
+            final DetailAST next = lastComma.getNextSibling();
+            final int type = next.getType();
+
+            if (type == TokenTypes.SEMI) {
+                if (next.getLineNo() == lastComma.getLineNo()) {
+                    // semi on the same line as last comma
+                    log(next.getLineNo(), MSG_KEY_SEMI);
+                }
+
+            }
+            else if (type == TokenTypes.ENUM_CONSTANT_DEF) {
+                log(next.getLineNo(), MSG_KEY);
+            }
```

Comment:
code should be like:
```
if ( isSmthAfterComm(lastComma, nextAst) ){
   log(next.getLineNo(), MSG_KEY);
}
```

Reply:
We need to distinguish between

``` java
    enum E1 {
        ONE,
        TWO,
        THREE,
    }
```

and

``` java
    enum E4 {
        ONE,
        TWO,
        THREE,; // violation
    }
```

In addition, there is not much that can come after a comma that follows an enum constant definition: another definition, or a semi colon.

Reply:
>Enum constant should contain trailing comma.
 vs
Enum constant should contain trailing comma without semi-colon afterwards.

"Enum constant should contain trailing comma"  can we just state that it should be the last symbol in line of code ?
@rnveach , what do you think ?

Reply:
Going over this message again, after more than two years (af7482f459b006813717399616212aaaa0752a3d), I think the message is not correct.

> Enum constant should contain trailing comma.

Better: "Enum constant should _be followed_ by a trailing comma."

Then, the second, more detailed message makes more sense too:

"Enum constant should _be followed_ by a trailing comma without semi-colon afterwards."

What do you think?

Reply:
@romani 
> "Enum constant should contain trailing comma" can we just state that it should be the last symbol in line of code ?

Wouldn't the `/` of a comment be a symbol? Check isn't comment aware, so it wouldn't throw a violation if a comment appeared after the comma.

Reply:
@romani 

> "Enum constant should contain trailing comma" can we just state that it should be the last symbol in line of code ?

Wouldn't the / of a comment be a symbol? Check isn't comment aware, so it wouldn't throw a violation if a comment appeared after the comma.

Reply:
In my last comment, I have suggested improved error messages. What about these?

Reply:
@kariem , @rnveach 
What about `Enum constant should be followed by a trailing comma only in line` ?

Reply:
The correct wording is `Enum constant should be followed by only a trailing comma in the line`.

@kariem I think @romani is trying to condense the 2 messages into 1.

Reply:
@rnveach @romani 

One message is definitely easier to implement. The thinking behind the current implementation was that we show the user the simplest message (`Enum constant should be followed by a trailing comma.`), and, only if necessary, tell them, that a semicolon on the same line is not allowed (`Enum constant should be followed by a trailing comma without semi-colon afterwards`).

Anyways, after your suggestions, how about the following:

`Enum constant should be followed by a trailing comma in the line.`

This the main point of the rule we are trying to enforce. I also think this message is very clear. In addition, if someone puts a semicolon right after the comma, we can show another message for clarity:

`Enum constant should be followed by a trailing comma in the line and no additional semicolon.`

The difference is that this is good

    enum E {
        ONE,
        TWO, // some comment
        THREE,
    }

while this is bad

    enum E {
        ONE,
        TWO,
        THREE,;
    }

The first allows for better diffs when adding or removing an enum constant, while the latter does not. This is why we should not write "by _only_ a trailing comma". It's not correct, as a comment is also fine (depending on other rules regarding documentation).

What do you think?

Reply:
> while the latter does not. 

Second code is also not good for diff.
It should be:
```
enum E {
    ONE,
    TWO,
    THREE,
     ;
}
```

That is why I am looking at single message. As we validating single and simple pattern - comma is after value, and last symbol at the line.

Am I missing smth ?

Reply:
> Second code is also not good for diff.

Yes, this is what I wrote.

The important feature is that you only change lines of code that are important for the change. If I add or remove one enum constant, only one line should be affected.

I understand the demand for a single message. I wanted to have two messages, one simple and one specific for the two cases. It's clear enough to write that there should be a trailing comma after every constant. If someone puts a semi-colon after the comma on the same line, then we should show the more specific message. If you think, only having one very specific message in there, please do not forget that specificity does not necessarily provide clarity. I always prefer the latter over the prior when we talk about error messages.

I also do not see a reason not to provide concise information for the user so that they know how to fix the violation. Is there a reason?

In addition, I do not think we should use the word "symbol" here. Especially as this code tries to enforce a level of consistency right above lexical correctness, we should not be to lax and use the correct terms. A comment is also a symbol, and both, traditional comments and end-of-line comments should not be considered a problem in the context of the check for trailing commas on enumerations. Hence, we should not use "last symbol" in the line, as there are actually two permitted symbols after the comma.

``` java
enum E {
    ONE,
    TWO,
    THREE, // this is ok
}
```

``` java
enum E {
    ONE,
    TWO, /* although confusing, this is also ok */
    THREE, 
}
```

``` java
enum E {
    ONE, /* although this does not make any sense, */ // this is also ok
    TWO, 
    THREE, 
}
```

@rnveach @romani, Please decide on either using one very specific message, or two concise and clear messages.

One message:

    Enum constant should be followed by only a trailing comma in the line.

We will still have to work on the wording, because, as you see in the examples above, this is not correct when we have comments on the same line after the enum constant (and trailing comma).

Two messages:

    Enum constant should be followed by a trailing comma in the line.
    Enum constant should be followed by a trailing comma in the line and no additional semicolon.

Reply:
> I also do not see a reason not to provide concise information for the user so that they know how to fix the violation. Is there a reason?

sorry that I did not shared a reason ... in sevntu project it is not a issue. But ones your Check start its way to main library .... each message need to be translated, each plugin need to probably register it and so on, so we try to avoid extra messages.

Looks like we need to vote. As there are three of us. 
@rnveach , please share your vote, and looks like your vote actually made a decision :).

> We will still have to work on the wording, because, as you see in the examples above, this is not correct when we have comments on the same line after the enum constant

I do not see big problem with this, almost all checks ignore comments completely.  I like your message.

Reply:
Done: 6c53368c49d477916945c1e69ee3c6f50546f78e

Reply:
Also, I have extracted the actual check to a separate method, as [previously suggested by @romani](https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/709#discussion_r217871003).